### PR TITLE
Feature/k8s spec native - part 1

### PR DESCRIPTION
**Pull request by ycliuhw**
### Checklist

 - [ ] ~Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?~
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

Add two new hook commands: `k8s-raw-set` and `k8s-raw-get`;

TODO: k8s provider is not implemented to handle raw spec yet.

## QA steps

In k8s charm, run `k8s-raw-set` instead of `pod_spec_set`;

```python
# This helper function will be added to CAAS layer soon.

from subprocess import run, PIPE, CalledProcessError

def k8s_raw_set(spec):
    try:
        run(['k8s-raw-set'], stdout=PIPE, stderr=PIPE, check=True, input=spec.encode('utf-8'))
    except CalledProcessError as err:
        stderr = err.stderr.decode('utf-8').strip()
        log(f'k8s-raw-set encountered an error: `{stderr}`', level='ERROR')

        if re.match(r'^ERROR application [\w-]+ not alive$', stderr):
            log('Ignored error due to k8s-raw-set getting called during app removal', level='INFO')
            return

        raise
```

```console
$ juju bootstrap microk8s k1

$ juju controller-config features='[raw-k8s-spec]'

$ juju add-model t1 microk8s

$ juju deploy /tmp/charm-builds/mariadb-k8s/ --debug  --resource mysql_image=mariadb

$ juju debug-log -m k1:controller --color --replay --level WARNING | grep 'raw k8s spec not implemented'
controller-0: 00:27:43 ERROR juju.worker.dependency "caas-unit-provisioner" manifold worker returned unexpected error: raw k8s spec not implemented

## login to Mongo;
$ juju:PRIMARY> use juju
switched to db juju
$ juju:PRIMARY> db.podSpecs.find().pretty()
{
	"_id" : "e3985c33-1263-46fb-8cf5-7d1cbb441997:a#mariadb-k8s",
	"spec" : "",
	"raw-spec" : "this is k8s raw spec v2!!!",
	"upgrade-counter" : 0,
	"model-uuid" : "e3985c33-1263-46fb-8cf5-7d1cbb441997",
	"txn-revno" : NumberLong(2),
	"txn-queue" : [
		"5e8c2b6ffe5c050019d5b03d_05a047b0"
	]
}

```

## Documentation changes

Yes

## Bug reference

None

### CODE REVIEW

**Comment by ycliuhw** [comment]
[k8s native yaml](https://trello.com/c/SJHeatKw/1087-k8s-native-yaml)

File: worker/uniter/runner/context/context.go [hunk 1]
```diff
@@ -1114,6 +1131,36 @@ func (ctx *HookContext) doFlush(process string) error {
...
+// If we're running the upgrade-charm hook and no podspec update was done,
+// we'll still trigger a change to a counter on the podspec so that we can
+// ensure any other charm changes (eg storage) are acted on.
+func (ctx *HookContext) addCommitHookChangesForCAAS(builder *uniter.CommitHookParamsBuilder, process string) error {
+	if ctx.podSpecYaml == nil && ctx.k8sRawSpecYaml == nil && process != string(hooks.UpgradeCharm) {
+		// No ops for any situation unless any k8s spec needs to be set or "upgrade-charm" was run.
+		return nil
+	}
+	if ctx.podSpecYaml != nil && ctx.k8sRawSpecYaml != nil {
+		return errors.NewForbidden(nil, "either k8s-spec-set or k8s-raw-set can be run for each application")
```
Comment (wallyworld) [inline worker/uniter/runner/context/context.go:1143]: ... for each application, but not both

File: apiserver/params/internal.go [hunk 2]
```diff
@@ -311,6 +311,7 @@ type CommitHookChangesArg struct {
 	SetUnitState         *SetUnitStateArg       `json:"unit-state,omitempty"`
 	AddStorage           []StorageAddParams     `json:"add-storage,omitempty"`
 	SetPodSpec           *PodSpec               `json:"pod-spec,omitempty"`
+	SetRawK8sSpec        *PodSpec               `json:"set-raw-k8s-spec,omitempty"`
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/params/internal.go:314]: After discussing, the existing approach is fine to start with

File: state/podspec_test.go [hunk 3]
```diff
@@ -42,14 +43,72 @@ func (s *PodSpecSuite) SetUpTest(c *gc.C) {
...
+	modelOp := s.Model.SetRawK8sSpecOperation(token, appTag, spec)
+	return s.State.ApplyOperation(modelOp)
+}
+
+func (s *PodSpecSuite) assertRawK8sSpecNotFound(c *gc.C, tag names.ApplicationTag) {
+	_, err := s.Model.RawK8sSpec(tag)
+	c.Assert(err, jc.Satisfies, errors.IsNotFound)
+}
+
+func (s *PodSpecSuite) TestSetRawK8sSpecOperationApplication(c *gc.C) {
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply state/podspec_test.go:70]: Yes, we have tests in TestWatchRawK8sSpec and TestWatchPodSpec to test set nil k8sspec and nil rawk8sspec.

File: worker/caasunitprovisioner/deployment_worker.go [hunk 4]
```diff
@@ -129,6 +129,14 @@ func (w *deploymentWorker) loop() error {
 		} else if err != nil {
 			return errors.Trace(err)
 		}
+		if info.RawK8sSpec != "" {
+			// TODO(caas): nothing we can do here before k8s provider can handle raw spec.
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply worker/caasunitprovisioner/deployment_worker.go:133]: So here will be removed soon in the next PR for making the k8s provider to handle raw k8s spec.

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
11:32:11 FAIL: worker_test.go:262: WorkerSuite.TestControllerConfigPubsubChange
$$merge$$

**Pull request merged**